### ISSUE TITLE
[Feature]: Aider-inspired RepoMap

### ISSUE THREAD

**Issue by neubig** [All-Hands-AI/OpenHands#2185]
**What problem or use case are you trying to solve?**

Aider has a functionality to create a [RepoMap](https://github.com/paul-gauthier/aider/blob/main/docs/repomap.md), which is a concise description of the repo in text format, with the most relevant/important parts highlighted.

**Describe the UX of the solution you'd like**

It would be nice to have a RepoMap class within OpenDevin that can be used by any agent to pull in a description of the repo.

**Do you have thoughts on the technical implementation?**

- [ ] copy the aider [repomap.py](https://github.com/paul-gauthier/aider/blob/main/aider/repomap.py) into OpenDevin. This could probably be put in either a new `indexing` folder [here](https://github.com/OpenDevin/OpenDevin/tree/main/opendevin), or in the [memory](https://github.com/OpenDevin/OpenDevin/tree/main/opendevin/memory) folder.
- [ ] make it possible to import the repomap into one of our competitive agents, such as CodeAct. This functionality should be optional, like the github message ([details](https://github.com/OpenDevin/OpenDevin/blob/f9c7c3a5205d294506a7b190b7621c0216239a3b/agenthub/codeact_agent/codeact_agent.py#L111-L114)).
- [ ] run experiments on SWE-Bench to see if this functionality improves accuracy

**Describe alternatives you've considered**

We could also implement this from scratch, or create improved code search functionality.

**Additional context**

* Parent issue #120

**Comment by neubig** [comment]
@ryanhoangt will probably take a look at this

**Comment by tobitege** [comment]
Reference: [#2248 Add Aider-inspired RepoMap](https://github.com/OpenDevin/OpenDevin/pull/2248)

**Comment by 0xdevalias** [comment]
See also:

- #742

**Comment by 0xdevalias** [comment]
IMO this shouldn't be closed as stale

**Comment by FellowTraveler** [comment]
repomap and graphrag are critical features IMO for coding agents.

**Comment by neubig** [comment]
Hey @TomLucidor, sorry about the trouble!

If you have a large repo OpenHands should not hang, although it might become confused. The hanging behavior is likely due to something else. We'd be happy to help diagnose this, but in order to do so we'll need to be able to reproduce the problem. The easiest way for us to do this is if you can press the ["feedback" button](https://docs.all-hands.dev/modules/usage/feedback) when you're encountering this behavior, or share logs. If you can do that it'd be great if you can open a new issue. I'd also be happy to discuss on [slack](https://join.slack.com/t/openhands-ai/shared_invite/zt-2tom0er4l-JeNUGHt_AxpEfIBstbLPiw).

**Comment by xingyaoww** [comment]
Hey @ryanhoangt can you self-assign this one?

**Comment by ryanhoangt** [comment]
Sure!

**Comment by BradKML** [comment]
As a little heads up, has anyone checked these repos out? (something along the lines of "codebase prompt")
- https://github.com/cyclotruc/gitingest
- https://github.com/mufeedvh/code2prompt
- https://github.com/raphaelmansuy/code2prompt
- https://github.com/kamilstanuch/codebase-digest

**Comment by enyst** [comment]
I found this kind of tool helpful, but not for feeding it as such to `openhands`, more for local work in or alongside vscode/cursor. It works for large codebases, to dump into Gemini, and go.

It's not great (Gemini is not great, at least pre-2.0), but I get something out of it. For now, I doubt it has the precision to be useful for the agent, considering the tradeoffs here. Have you tried to give it like that, in a message?

**Comment by BradKML** [comment]
@enyst what do you have in mind compared to the established tools made specifically for chatbots? Cus I am tempted but not even sure whether to put/generate it in a text file for the repo, OR to throw it directly in the dialogue box. Asking second opinions

**Comment by ryanpeach** [comment]
Just wanted to point out, you can import `aider.repomap` and get this feature for free.

**Comment by ryanpeach** [comment]
@rbren where is the PR that completed this?

**Comment by raymyers** [comment]
Just noting here that @xingyaoww discussed the reason for closing this issue in today's community meetup (Jan 31, 2025), so until more info is added here, it should be in the recording about 15 min in.

**Comment by xingyaoww** [comment]
I believe @ryanhoangt can explain more about this, but in summary, we've been trying something like this: https://github.com/All-Hands-AI/openhands-aci/pull/41 for integrating Aider-inspired Repo map -- and it seems we didn't get it working on claude after a few attempts (e.g., performance on SWE-Bench didn't improve when adding those file skeleton) -- hoang feel free to correct me if I'm wrong!

> I found this kind of tool helpful, but not for feeding it as such to openhands, more for local work in or alongside vscode/cursor. It works for large codebases, to dump into Gemini, and go. 

As Engel said, there's a significant trade-off here:
* For single-turn systems where the LLM typically doesn't have the autonomy to look around and execute actions (e.g., I need to paste context to ChatGPT/Claude for it to generate another block of code), such type of RepoMap is essential -- an example of this would be Aider that primarily focuses on "file-editing" action, as opposed to executing general bash command like "grep -n".
* However, for a flexible multi-turn agentic system like OpenHands (CodeAct) -- sometimes adding more of this info actually did more harm than good. Basically, if you don't tell the agent anything, it would think, "I need to poke around and figure out right context,"; but if you tell it something like RepoMap (but it could be incomplete!) -- the agent would trust your RepoMap and think that's everything they need to know to complete the task. Here's another bitter lesson that happened recently (https://github.com/All-Hands-AI/OpenHands/pull/6491): we *think* providing less context of the file tree structure to the agent would help, but it actually hurt things a lot - so definitely, "what humans think is useful might not hold true for autonomous multi-turn agents"!

I would think things like RepoMap were initially created to create a context for a "single-turn LLM-based system." 
However, they may not be useful for multi-turn systems like OpenHands, because **the agent has the autonomy to look for stuff they need** -- and a lot of times, it did better than "our human-curated repo map based on our intuition of what the model needs" -- another vivid reminder of the [bitter lesson](http://www.incompleteideas.net/IncIdeas/BitterLesson.html) 😢 .

Here are some baseline code localization performance numbers on a paper I've been working on with some folks:

![Image](https://github.com/user-attachments/assets/18d4705f-1059-49e0-be8d-a22497a21eaa)

You can see even if we don't have any system like "RepoMap" to manually construct context within a large repo - whereas another system like "agentless" and "moatless" all have RepoMap-style systems didn't really outperform OpenHands on everything (if not worse!).

Based on these existing results, I've personally feel it is probably less productive to keep pushing towards this direction compared to other things after spending a lot of effort here -- but I could be totally wrong, and would be open to re-open this PR if people from the community is interested in taking over and actually making it work ❤

**Comment by xingyaoww** [comment]
cc @jimwhite -- thanks for the great discussion in today's community meeting! 👆 hopefully this response above would better answer your question :)

**Comment by aymannadeem** [comment]
Hey @xingyaoww! You mentioned that agents often do better when left to explore versus being given a repo map. Could you share some specific examples where the agent's autonomous exploration led to better solutions than when provided with structured context?

**Comment by BradKML** [comment]
I am thinking something like `eza` can make dynamic folder maps https://github.com/eza-community/eza
Right now when I am testing OH often times the agent would forget the `pwd` or context of where the terminal is, after `cd`ing in or out, which is annoying and warrant some kind of tool that shows both `ls` and hints at `pwd`.

**Comment by enyst** [comment]
> often times the agent would forget the `pwd` or context of where the terminal is, after `cd`ing in or out, which is annoying and warrant some kind of tool that shows both `ls` and hints at `pwd`.

The LLM is already supposed to get in the prompt, after each command, the current directory:
https://github.com/All-Hands-AI/OpenHands/blob/fe8b92743bf13676b331209547685e22062596f7/openhands/events/observation/commands.py#L159

If that is not happening, maybe you could post a new issue, with some log with it missing or wrong? But I think it's also possible that we send it and sometimes the LLM ignores it anyway (depending on LLM).

**Comment by BradKML** [comment]
It might be an LLM issue then @enyst since I am heavily leaning on DeepSeek v3 atm, but it can't invoke the observation whenever it moves around working directories it seems?

**Comment by enyst** [comment]
It's automatic. The LLM asks for a command to execute ('cd ...'), the environment performs it and responds with an observation which includes `cwd` as above.

**Comment by ryanpeach** [comment]
Very interesting writeup @xingyaoww

**Comment by xingyaoww** [comment]
As promised! I'm sharing official result above and our preprint here: https://arxiv.org/abs/2503.09089

[@Zhaoling Chen](https://openhands-ai.slack.com/team/U07A8PMDGK0), [@Robert Tang](https://openhands-ai.slack.com/team/U06PJ6EKXGF), @Gangda Deng and I were working to figure out localization problems in coding agents using OpenHands :thinking_face: Finally be able to share the pre-print:
- We build a graph-based repository indexing (that only requires static analyzer)
  - Allow LLMs to perform complex multi-hop navigation & reasoning
  - only takes a few seconds per codebase (faster than RAG!)
- LLM as agents equipped with this indexing perform better than the existing tools we have in openhands (e.g., grep via bash)
- And most interesting to me, we finally show that frontier model + agent can outperform plain embedding-based or RAG-based procedure (Agentless/Moatless)
  - This probably tells us "a strong agent + graph-based indexing is all you need". There's no need to construct the embedding for the ENTIRE repo for retrieval -- it adds cost of embedding & a lot of engineering challenges: you need to update the embedding when the codebase is updated; you need to wait for the codebase to be indexed when you first open a new repo
  - The only downside here is that running these strong agents are expensive
  - :point_up_2: BUT! We were able to train a Qwen 2.5 32B model that almost perform on-par as Claude-3.5 -- and the fine-tuned 7B model is already competitive with OpenHands + sonnet 3.5 in function-level Acc@10. This should greatly help with the cost

[@Zhaoling Chen](https://openhands-ai.slack.com/team/U07A8PMDGK0) will help integrate this LocAgent into OpenHands in the coming weeks so stay tuned :slightly_smiling_face:

![Image](https://github.com/user-attachments/assets/7015c33c-01f3-4f4b-9d45-aa6e173bb1f6)

![Image](https://github.com/user-attachments/assets/ebcffe2d-f02f-44cd-a7de-274bec6ab28f)

**Comment by paul-gauthier** [comment]
Looks like an interesting paper. Thanks for the somewhat oblique and generic citation to aider.

Is there a reason you didn't directly cite aider's repo-map, discuss it in the related work section or include it in your experimental evaluations?

At a *minimum*, your work seems quite strongly "inspired by" aider's repo-map. You've been participating in this GitHub issue titled "Aider-inspired RepoMap" for months, discussing your attempts to implement and evaluate a version of it in OpenHands. 

Your top-listed contribution sounds very familiar.

> Our contributions address critical gaps in existing approaches: We introduce a heterogeneous graph representation that captures both explicit and implicit code relationships, enabling efficient multi-hop reasoning. Our lightweight graph-based indexing process takes only seconds per repository and requires minimal storage.

Aider's original repo-map code and blog article were published in May 2023. A follow up blog article was published in Oct 2023, describing the tree-sitter implementation of repo-map.

**Comment by jmvldz** [comment]
Very exciting @xingyaoww! Thank you for sharing and for the follow up on this thread.

**Comment by xingyaoww** [comment]
@paul-gauthier Sorry for the oversight & appreciate the feedback! I didn't realize there's a specific blog directly discussing the "repo-map"- we will be sure to cite and discuss them in the next revision of the preprint :) (cc @czlll)

I have to admit, even though we are discussing this under the thread of Aider-inspired RepoMap, the question we were trying to answer in this paper is more about which is better, "agentic-based information gathering" or "passively provided information" (e.g., via embedding, where the LLM has few options to control the information being retrieved). IRRC, the original AiderMap is designed primarily for the latter category, and it didn't immediately jump into my mind when reviewing the paper 😅 .

But now, reflecting back, the indexing approach we took does look similar to what AiderMap was doing under the hood and we will certainly discuss it and give credits.

I'd say the differences between LocAgent vs. Aider are more about "how LLMs are using these graph-based indexing." My (maybe outdated) knowledge of Aider is that it automatically constructs a RepoMap for LLM (based on previously mentioned keywords, etc) to provide the context for LLM to perform edit (more similar to the passively provided information); whereas LocAgent is more about allowing LLM to use tools to explicitly traverse the graph (more leaning towards agentic-based information gathering - which is something we are hoping to continue to do in OpenHands).

**Comment by tangxiangru** [comment]
@paul-gauthier We will cite aider's repo-map in our next version. Thanks.
